src/backend/distributed/metadata/metadata_sync.c

```diff
@@ -0,0 +1,474 @@
+					 " nodename, nodeport) "
+					 "VALUES ");
+
+	/* add placements to insertPlacementCommand */
+	foreach(shardIntervalCell, shardIntervalList)
+	{
+		ShardInterval *shardInterval = (ShardInterval *) lfirst(shardIntervalCell);
+		uint64 shardId = shardInterval->shardId;
+
+		List *shardPlacementList = FinalizedShardPlacementList(shardId);
+		ShardPlacement *placement = NULL;
+
+		/* the function only handles single placement per shard */
+		Assert(list_length(shardPlacementList) == 1);
+
+		placement = (ShardPlacement *) linitial(shardPlacementList);
+
+		appendStringInfo(insertPlacementCommand,
+						 "(%lu, 1, %lu, %s, %d)",
+						 shardId,
+						 placement->shardLength,
+						 quote_literal_cstr(placement->nodeName),
+						 placement->nodePort);
+
+		processedShardPlacementCount++;
+		if (processedShardPlacementCount != shardCount)
+		{
+			appendStringInfo(insertPlacementCommand, ",");
+		}
+	}
+
+	/* add the command to the list that we'll return */
+	commandList = lappend(commandList, insertPlacementCommand->data);
+
+	/* now, generate the shard query without any values yet */
+	appendStringInfo(insertShardCommand,
+					 "INSERT INTO pg_dist_shard "
+					 "(logicalrelid, shardid, shardstorage,"
+					 " shardminvalue, shardmaxvalue) "
+					 "VALUES ");
+
+	/* now add shards to insertShardCommand */
+	foreach(shardIntervalCell, shardIntervalList)
+	{
+		ShardInterval *shardInterval = (ShardInterval *) lfirst(shardIntervalCell);
+		uint64 shardId = shardInterval->shardId;
+		Oid distributedRelationId = shardInterval->relationId;
+		char *qualifiedRelationName = generate_qualified_relation_name(
+			distributedRelationId);
+
+		/* TODO: consider other types */
+		int minHashToken = DatumGetInt32(shardInterval->minValue);
+		int maxHashToken = DatumGetInt32(shardInterval->maxValue);
+
+		appendStringInfo(insertShardCommand,
+						 "(%s::regclass, %lu, '%c', '%d', '%d')",
+						 quote_literal_cstr(qualifiedRelationName),
+						 shardId,
+						 shardInterval->storageType,
+						 minHashToken,
+						 maxHashToken);
+
+		processedShardCount++;
+		if (processedShardCount != shardCount)
+		{
+			appendStringInfo(insertShardCommand, ",");
+		}
+	}
+
+	/* finally add the command to the list that we'll return */
+	commandList = lappend(commandList, insertShardCommand->data);
+
+	return commandList;
+}
+
+
+/*
+ * worker_drop_distributed_table drops the distributed table with the given oid,
+ * then, removes the associated rows from pg_dist_partition, pg_dist_shard and
+ * pg_dist_shard_placement. The function also drops the server for foreign tables.
+ *
+ * Note that drop fails if any dependent objects are present for any of the
+ * distributed tables. Also, shard placements of the distributed tables are
+ * not dropped as in the case of "DROP TABLE distributed_table;" command.
+ *
+ * The function errors out if the input relation Oid is not a regular or foreign table.
+ */
+Datum
+worker_drop_distributed_table(PG_FUNCTION_ARGS)
```

Comment:
It might make sense to put this function in a different file, e.g. worker/worker_drop_protocol.c.